src/zjs_webusb.c

```diff
@@ -217,28 +241,343 @@ static ZJS_DECL_FUNC(zjs_webusb_set_url)
     return ZJS_UNDEFINED;
 }
 
+int fill_uart(struct device *dev, ringbuf_t *ring, int length)
+{
+    // requires: dev is a UART device, ring is a ring buffer, length is the max
+    //             bytes to read, length must not be enough to overflow buffer
+    //  returns: total number of bytes sent
+    //  effects: fill UART dev with up to length bytes from ring at index, until
+    //             the fifo is full
+    int total = 0;
+    while (length > 0) {
+        int index = ring->read;
+        int sent = uart_fifo_fill(dev, &ring->buf[index], length);
+        if (sent == 0) {
+            break;
+        }
+        index += sent;
+        total += sent;
+        if (index == RINGMAX) {
+            index = 0;
+        }
+        ring->read = index;
+        if (sent == length) {
+            break;
+        }
+        length -= sent;
+    }
+    return total;
+}
+
+int read_uart(struct device *dev, ringbuf_t *ring, int length)
+{
+    // requires: dev is a UART device, ring is a ring buffer, length is the max
+    //             bytes to write, length must not be enough to overflow buffer
+    //  returns: true if all bytes read, false if more remain
+    //  effects: consume bytes from UART dev up to length bytes into ring at
+    //             index, until the fifo runs out
+    int total = 0;
+    while (length > 0) {
+        int index = ring->write;
+        char tmp[100];
+        int bytes = uart_fifo_read(dev, tmp, length);
+        memcpy(&ring->buf[index], tmp, bytes);
```

Comment:
can uart_fifo_read() return more than 100 bytes per read? possible overflow here

Reply:
Good catch, i think this was a partial attempt at an improvement... I don't really want to read in two chunks, so it would be better to read the whole thing and then write to end of ring buf, then wrap to beginning at once.

Reply:
Ok, add a FIXME here so we don't lose track of this